### PR TITLE
Enhancement/skale 3070 hashing features

### PR DESCRIPTION
**Pull request by payvint**
SKALE-3070 Implemented Hashing features

### CODE REVIEW

File: contracts/KeyStorage.sol [hunk 1]
```diff
@@ -42,44 +42,15 @@ contract KeyStorage is Permissions {
         bytes32 share;
     }
 
+    // Unused variable!!
```
Comment (cstrangedk) [inline contracts/KeyStorage.sol:45]: Why not remove?

File: contracts/SkaleDKG.sol [hunk 2]
```diff
@@ -134,13 +145,15 @@ contract SkaleDKG is Permissions, ISkaleDKG {
...
-
-        _isBroadcast(
-            groupIndex,
-            nodeIndex,
-            secretKeyContribution,
-            verificationVector
-        );
+        uint index = _nodeIndexInSchain(groupIndex, nodeIndex);
+        require(index < channels[groupIndex].n, "Node is not in this group");
+        require(!dkgProcess[groupIndex].broadcasted[index], "This node is already broadcasted");
```
Comment (cstrangedk) [inline contracts/SkaleDKG.sol]: This node has...

**Review by cstrangedk — CHANGES_REQUESTED** [review]

File: contracts/KeyStorage.sol [hunk 3]
```diff
@@ -268,70 +162,7 @@ contract KeyStorage is Permissions {
...
-        // Decrypt secret key contribution
-        SchainsInternal schainsInternal = SchainsInternal(contractManager.getContract("SchainsInternal"));
-        uint index = schainsInternal.getNodeIndexInGroup(groupIndex, fromNodeToComplaint);
-        uint indexOfNode = schainsInternal.getNodeIndexInGroup(groupIndex, nodeToComplaint);
-        uint secret = Decryption(contractManager.getContract("Decryption")).decrypt(
-            _data[groupIndex][indexOfNode].secretKeyContribution[index].share,
-            key
-        );
-        return secret;
+    function _getData() private view returns (BroadcastedData memory) {
```
Comment (olehnikolaiev) [inline contracts/KeyStorage.sol:165]: not needed

**Review by olehnikolaiev — CHANGES_REQUESTED** [review]

File: contracts/KeyStorage.sol (hunk 1 above)
Comment (payvint) [reply contracts/KeyStorage.sol:45]: We have already deployed contract to mainnet with this variable!

File: contracts/SkaleDKG.sol (hunk 2 above)
Comment (payvint) [reply contracts/SkaleDKG.sol]: Thanks:)

File: contracts/KeyStorage.sol (hunk 3 above)
Comment (payvint) [reply contracts/KeyStorage.sol:165]: Yeah not needed but if we leave _data (contract deployed) we need to use this variable somewhere and I have created unused function

**Review by cstrangedk — APPROVED** [review]
lgtm

**Pull request merged**